### ISSUE TITLE
ZeroTier Android breaks Play Store app updates [not our issue]

### ISSUE THREAD

**Issue by nolanl** [zerotier/ZeroTierOne#469]
On multiple Wifi networks (minimally one using 192.168.1.0/24, one using 192.168.0.0/24) Google Play Store times out attempting to update already-installed apps.  Installing new apps from the store also appears to be affected.  I've also seen the issue with no wifi, only cellular network (using real public IPs).

The Private network in question (the only one on the device) uses 192.168.192.0/24.  Disconnecting from that network in the ZeroTier app does _not_ fix the issue.  Disconnecting the ZeroTier One VPN from Android's VPN settings page does.  Connecting to another VPN (L2TP) on the Android VPN setting page does not cause the issue.

I've not noticed any connectivity problems with other apps, just Google Play Store.

**Comment by nolanl** [comment]
Android 7.1.1 on a Nexus 5X.

Patch level: March 5th, 2017, though it has also happened on older patch levels.

**Comment by nolanl** [comment]
Latest version of ZeroTier (1.2.2-1), though it also happened on older versions.

**Comment by jefft** [comment]
Confirmed.

**Comment by Arffeh** [comment]
Download manager refuses to download as well. So the internal application
for downloads from apps and the play store is probably seeing an active VPN
and attempting to use it.

**Comment by adamierymenko** [comment]
So this happens even if we don't set a default route? That's stupid. Not setting a default route means this is *not* a full-tunnel VPN.

**Comment by Arffeh** [comment]
One other thing I've noticed is that if you go into the ZT1 application and
disconnect entirely, it no longer closes the android VPN tunnel interface.
You have to go into androids VPN manager and kill it there too.

I skimmed the android VPN documentation, there is mention of the disconnect
call needed to remove the tunnel entirely.

**Comment by glimberg** [comment]
@Arffeh I've pushed out a beta release `1.2.2-2` that fixes the disconnect button not properly tearing down the VPN interface.  If you'rea not already in the beta stream, you can join it here:  https://play.google.com/apps/testing/com.zerotier.one

As for the other issues in this thread, thus far it appears to be an Android problem.  Unless you have a default route configured on your network on https://my.zerotier.com *AND* you have `Route via ZeroTier` enabled in the app, there is no default route set for the VPN interface.  So unless Google's IP blocks are somehow falling into the configured routes for your network(s), the issues above should not be happening.

**Comment by Arffeh** [comment]
(Confirmed, full teardown works on the beta.)


Looking through the android bug tracker, it seems there are similar symptoms, but nothing exact.
Of note, https://code.google.com/p/android/issues/detail?id=233963
Their tunnel only has IPv6 connectivity, and magically their download manager stops using the main connection.

Should we put up a bug on their tracker for our use case?

**Comment by adamierymenko** [comment]
This is not the first time we've found a bug in an OS. (Windows hint hint Windows.)

So this is fixed / not ours?

**Comment by nolanl** [comment]
Still happens with 7.1.2.

Hopefully a workaround is possible, because this makes ZeroTier pretty hard to use on Android.  Even if they fix the underlying bug upstream, that won't help the majority of people on crappy vendor handsets for many years...

**Comment by Arffeh** [comment]
Adam, I feel like this isn't your issue. Unfortunately that means getting it fixed would be a nightmare.
It's been the case since android 7.* as far as I can tell. I first noticed this when going from 6 to 7, but dismissed it and worked around it by disabling ZT1 every time I needed something.
I'm guessing in their bid to try and get a bit more security in their OS, they've accidentally forced Download Manager to blindly use VPNs if they are active, regardless of the state of the provided VPN.
This behaviour flows down to any application using the DownloadManager API to provoke downloads, which happens to be a lot of applications.

Even if it doesn't help "people on crappy vendor handsets" who can't/don't want to use something with nightly/rolling releases, submitting a bug and getting it corrected for the people who use up to date ROMs would be extremely beneficial for the appearance of ZT1. Even though it's an android bug, ignoring it means that ultimately ZT1 looks "broken". 

To give you an idea of what getting it corrected upstream, just LineageOS alone has [nearly 1.5 million active installs](https://stats.lineageos.org/), and it's only been properly downloadable since Feb 10th. Not exactly a small audience.

**Comment by nolanl** [comment]
Looks like it is pretty well known since the N dev previews:
https://issuetracker.google.com/issues/37105245
https://issuetracker.google.com/issues/37099545

**Comment by glimberg** [comment]
We've gotten more reports from users that this doesn't just affect the Play store app.  It appeared to be happening on *all* Google branded apps (Chrome, Play Store, etc).  We're still looking for a full solution if we can find one, but so far we can't even recreate this issue on any of the Android devices we have here at the office.  It appears to be some sort of nastiness in the OS when adding IPv6 to ZeroTier networks in some network environments.

In the mean time, I've just pushed out another release to the Beta channel that allows users to disable the use of IPv6 on ZeroTier networks.

**Comment by nolanl** [comment]
Confirmed: using your beta, and disabling IPv6 fixes the issue.  Thanks for the workaround!  Not your bug, but totally a shame...

**Comment by glimberg** [comment]
Glad it works for you! I'm still trying to find a way to reliably reproduce it once while at home.  But by the time I got to the computer and had the debugger attached, everything worked just fine again

**Comment by nolanl** [comment]
Random thought: since ZeroTier is creating an L2 network, could the problem be an RA leaking in from a different host on the zt network?  That might explain why it doesn't always show up; it would depend on the other hosts on the zt network...

**Comment by glimberg** [comment]
Not sure on that.  All I know is that ZeroTier isn't the only app that uses Android's VPN subsystem having this issue.

**Comment by HarukaMa** [comment]
Sorry for reviving this issue, but I'm having similar problems with Play Store. I don't have IPv6 configured on ZT, and have set default route in controller and enabled it on device (0.0.0.0/0 via 11.0.0.1 and 11.0.0.0/8 via null). 

I have tried beta version and disabled IPv6 but Play Store is still not working. (Also in the beta version, "Broadcast" is always false even though I have set `"enableBroadcast": true`.)

I'm unclear if the connection went through wifi or couldn't connect at all, as I can't connect to Play Store without something like VPN.

Android 7.0 and bl not unlockable (unrooted) so I couldn't get more information out of my device...
